src/lib/AudioUltra/Media/AudioDecoder.ts

```diff
@@ -0,0 +1,260 @@
+import { AudioDecoderWorker, getAudioDecoderWorker } from 'audio-file-decoder';
+// eslint-disable-next-line
+// @ts-ignore
+import DecodeAudioWasm from 'audio-file-decoder/decode-audio.wasm';
+import { Events } from '../Common/Events';
+import { info } from '../Common/Utils';
+
+const DURATION_CHUNK_SIZE = 60 * 30; // 30 minutes
+
+export const DEFAULT_FREQUENCY_HZ = 44100;
+
+interface AudioDecoderEvents {
+  progress: (chunk: number, total: number) => void;
+}
+
+export class AudioDecoder extends Events<AudioDecoderEvents> {
+  static cache: Map<string, AudioDecoder> = new Map();
+
+  chunks?: Float32Array[];
+  private cancelled = false;
+  private decodeId = 0; // if id=0, decode is not in progress
+  private worker: AudioDecoderWorker | undefined;
+  private decoderResolve?: () => void;
+  private decodingResolve?: () => void;
+  private _dataLength = 0;
+  private _dataSize = 0;
+  private _channelCount = 1;
+  private _sampleRate = DEFAULT_FREQUENCY_HZ;
+  private _duration = 0;
+  decoderPromise: Promise<void> | undefined;
+  decodingPromise: Promise<void> | undefined;
+
+  constructor(private src: string) {
+    // Return a cached instance if it exists
+    // This allows for multiple instances of the same audio file source to share the same worker
+    // and decoded data results
+    if (AudioDecoder.cache.has(src)) {
+      const instance = AudioDecoder.cache.get(src) as AudioDecoder;
+
+      instance.renew();
+
+      AudioDecoder.cache.set(src, instance);
+
+      return instance;
+    }
+
+    super();
+
+    // only allow one cached decoder at a time to prevent memory leaks
+    // and limit the memory usage of the browser
+    if (AudioDecoder.cache.size > 0) {
```

Comment:
Does it work well when there are two `Audio` tags in the config?

Reply:
Currently if they are different sources, it would cancel one the first decode result. I'm going to adjust this to instead cleanup instances on destroy with a buffer period, so it doesn't throw away in-use work (ex. switching annotations, submitting annotations, view all etc.), but allows for multiple decoder results to exist in parallel. This will not play nicely with very large audio, but will be fine for moderate sized durations.